panic: send on closed channel

## Packages

Scylla version: `2025.1.4-20250707.20afd2776561` with build-id `36ff3bad104de1121fdc101e72233d77111253c1`





Kernel Version: `6.8.0-1031-aws`


## Issue description

- [ ] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

*Describe your issue in detail and steps it took to produce it.*
gemini ran 8 hours, then got panic at test end:
```
2025-07-16 01:10:25.093: (GeminiStressEvent Severity.ERROR) period_type=end event_id=84081b53-411b-4096-94c8-2e6a306c06a1 duration=8h10m9s: node=Node gemini-tombstones-gemini-t-loader-node-e621bfbb-1 [3.253.99.129 | 10.4.3.193] (rack: RACK0) gemini_cmd=gemini                 --test-cluster="10.4.0.230,10.4.1.192,10.4.2.102,10.4.0.239,10.4.3.180,10.4.3.151"                 --seed=54                 --schema-seed=54                 --profiling-port=6060                 --bind=0.0.0.0:2112                 --outfile=/gemini_result_7d9b0dc6-4707-443c-ac52-2d9bd17cc4a1.log                 --replication-strategy="{'class': 'NetworkTopologyStrategy', 'replication_factor': '3'}"                 --oracle-replication-strategy="{'class': 'NetworkTopologyStrategy', 'replication_factor': '1'}" --oracle-cluster="10.4.0.241" --table-options="gc_grace_seconds=60 " --level=info --request-timeout=3s --connect-timeout=60s --consistency=QUORUM --async-objects-stabilization-backoff=10ms --async-objects-stabilization-attempts=10 --dataset-size=large --oracle-host-selection-policy=token-aware --test-host-selection-policy=token-aware --drop-schema=true --cql-features=normal --materialized-views=false --use-server-timestamps=true --use-lwt=false --use-counters=false --max-tables=1 --max-columns=16 --min-columns=8 --max-partition-keys=6 --min-partition-keys=2 --max-clustering-keys=4 --min-clustering-keys=2 --partition-key-distribution=uniform --partition-key-buffer-reuse-size=256 --statement-log-file-compression=zstd --io-worker-pool=2048 --duration 8h --warmup 10m --concurrency 100 --mode mixed --max-mutation-retries-backoff 10s --max-mutation-retries 60 --token-range-slices 20000 --max-errors-to-store 1
result=Exit code: 2
Command output: ['{"level":"info","ts":"2025-07-15T17:10:21.593101015Z","logger":"jobs","msg":"start jobs"}', '{"level":"info","ts":"2025-07-16T01:10:21.690790382Z","logger":"jobs","msg":"stop jobs"}']
errors=['Command error: panic: send on closed channel\n\ngoroutine 101 [running]:\ngithub.com/scylladb/gemini/pkg/generators.(*Partition).push(...)\n\t/home/runner/work/gemini/gemini/pkg/generators/partition.go:90\ngithub.com/scylladb/gemini/pkg/generators.(*Generator).fillAllPartitions.func2()\n\t/home/runner/work/gemini/gemini/pkg/generators/generator.go:254 +0x273\ngithub.com/scylladb/gemini/pkg/metrics.(*RunningTime).RunFuncE(0xc000105e68, 0xc000105ed0)\n\t/home/runner/work/gemini/gemini/pkg/metrics/metrics.go:290 +0x51\ngithub.com/scylladb/gemini/pkg/generators.(*Generator).fillAllPartitions(0xc0001382c0)\n\t/home/runner/work/gemini/gemini/pkg/generators/generator.go:240 +0x271\ngithub.com/scylladb/gemini/pkg/generators.(*Generator).start.func2()\n\t/home/runner/work/gemini/gemini/pkg/generators/generator.go:177 +0x45\ncreated by github.com/scylladb/gemini/pkg/generators.(*Generator).start in goroutine 1\n\t/home/runner/work/gemini/gemini/pkg/generators/generator.go:171 +0x16a\n\n']
```
tear down:
```
< t:2025-07-16 01:10:49,523 f:tester.py       l:3047 c:GeminiTest           p:INFO  > TearDown is starting...
```

gemini command:

```shell
gemini                 
--test-cluster="10.4.0.230,10.4.1.192,10.4.2.102,10.4.0.239,10.4.3.180,10.4.3.151"                 
--seed=54                 
--schema-seed=54
--profiling-port=6060                 
--bind=0.0.0.0:2112                 
--outfile=/gemini_result_7d9b0dc6-4707-443c-ac52-2d9bd17cc4a1.log                 
--replication-strategy="{'class': 'NetworkTopologyStrategy', 'replication_factor': '3'}"                 
--oracle-replication-strategy="{'class': 'NetworkTopologyStrategy', 'replication_factor': '1'}" 
--oracle-cluster="10.4.0.241" 
--table-options="gc_grace_seconds=60 " 
--level=info 
--request-timeout=3s 
--connect-timeout=60s 
--consistency=QUORUM 
--async-objects-stabilization-backoff=10ms 
--async-objects-stabilization-attempts=10 
--dataset-size=large 
--oracle-host-selection-policy=token-aware 
--test-host-selection-policy=token-aware 
--drop-schema=true 
--cql-features=normal 
--materialized-views=false 
--use-server-timestamps=true 
--use-lwt=false 
--use-counters=false 
--max-tables=1 
--max-columns=16 
--min-columns=8 
--max-partition-keys=6 
--min-partition-keys=2 
--max-clustering-keys=4 
--min-clustering-keys=2 
--partition-key-distribution=uniform 
--partition-key-buffer-reuse-size=256 
--statement-log-file-compression=zstd 
--io-worker-pool=2048 
--duration 8h 
--warmup 10m 
--concurrency 100 
--mode mixed 
--max-mutation-retries-backoff 10s 
--max-mutation-retries 60 
--token-range-slices 20000 
--max-errors-to-store 1
```

## Impact

*Describe the impact this issue causes to the user.*

## How frequently does it reproduce?

*Describe the frequency with how this issue can be reproduced.*

## Installation details

Cluster size: 6 nodes (i4i.2xlarge)

Scylla Nodes used in this run:
- gemini-tombstones-gemini-t-oracle-db-node-e621bfbb-1 (54.171.134.194 | 10.4.0.241) (shards: 30)
- gemini-tombstones-gemini-t-db-node-e621bfbb-6 (18.202.236.60 | 10.4.3.151) (shards: 7)
- gemini-tombstones-gemini-t-db-node-e621bfbb-5 (54.246.227.119 | 10.4.3.180) (shards: 7)
- gemini-tombstones-gemini-t-db-node-e621bfbb-4 (34.247.195.129 | 10.4.0.239) (shards: 7)
- gemini-tombstones-gemini-t-db-node-e621bfbb-3 (34.253.227.225 | 10.4.2.102) (shards: 7)
- gemini-tombstones-gemini-t-db-node-e621bfbb-2 (54.72.211.59 | 10.4.1.192) (shards: 7)
- gemini-tombstones-gemini-t-db-node-e621bfbb-1 (108.129.87.175 | 10.4.0.230) (shards: 7)


OS / Image: `ami-0252ea9d9736d854b` (aws: undefined_region)

Test: `gemini-tombstones`
Test id: `e621bfbb-00f9-4ca7-a37c-007fb1d84d1d`
Test name: `scylla-staging/yarongilor/gemini-tombstones`
Test method: `gemini_test.GeminiTest.test_load_random_with_nemesis`
Test config file(s):

- [gemini-tombstones.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/3c60e79ede36d88620bd82c15f3f7b49425bdabc/test-cases/gemini/gemini-tombstones.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor e621bfbb-00f9-4ca7-a37c-007fb1d84d1d`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=e621bfbb-00f9-4ca7-a37c-007fb1d84d1d)
- Show all stored logs command: `$ hydra investigate show-logs e621bfbb-00f9-4ca7-a37c-007fb1d84d1d`


## Logs:
- **db-cluster-e621bfbb.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/db-cluster-e621bfbb.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/db-cluster-e621bfbb.tar.zst)
- **schema-logs-e621bfbb.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/schema-logs-e621bfbb.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/schema-logs-e621bfbb.tar.zst)
- **sct-runner-events-e621bfbb.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/sct-runner-events-e621bfbb.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/sct-runner-events-e621bfbb.tar.zst)
- **sct-e621bfbb.log.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/sct-e621bfbb.log.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/sct-e621bfbb.log.tar.zst)
- **loader-set-e621bfbb.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/loader-set-e621bfbb.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/loader-set-e621bfbb.tar.zst)
- **monitor-set-e621bfbb.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/monitor-set-e621bfbb.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/20250716_011424/monitor-set-e621bfbb.tar.zst)
- **builder-e621bfbb.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/upload_20250716_011609/builder-e621bfbb.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/e621bfbb-00f9-4ca7-a37c-007fb1d84d1d/upload_20250716_011609/builder-e621bfbb.log.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/gemini-tombstones/37/)
[Argus](https://argus.scylladb.com/test/26378332-31c5-4d01-ad20-4dabfa8813ec/runs?additionalRuns[]=e621bfbb-00f9-4ca7-a37c-007fb1d84d1d)
</details>

a bit more readable output of the panic
```
panic: send on closed channel

goroutine 101 [running]:
github.com/scylladb/gemini/pkg/generators.(*Partition).push(...)
	/home/runner/work/gemini/gemini/pkg/generators/partition.go:90
github.com/scylladb/gemini/pkg/generators.(*Generator).fillAllPartitions.func2()
	/home/runner/work/gemini/gemini/pkg/generators/generator.go:254 +0x273
github.com/scylladb/gemini/pkg/metrics.(*RunningTime).RunFuncE(0xc000105e68, 0xc000105ed0)
	/home/runner/work/gemini/gemini/pkg/metrics/metrics.go:290 +0x51
github.com/scylladb/gemini/pkg/generators.(*Generator).fillAllPartitions(0xc0001382c0)
	/home/runner/work/gemini/gemini/pkg/generators/generator.go:240 +0x271
github.com/scylladb/gemini/pkg/generators.(*Generator).start.func2()
	/home/runner/work/gemini/gemini/pkg/generators/generator.go:177 +0x45
created by github.com/scylladb/gemini/pkg/generators.(*Generator).start in goroutine 1
	/home/runner/work/gemini/gemini/pkg/generators/generator.go:171 +0x16a
```

I would assume gemini has problem here with order of stopping things on shutdown.

--- a/pkg/generators/generator.go
+++ b/pkg/generators/generator.go
@@ -43,7 +43,6 @@ type Interface interface {
 	Get(context.Context) (typedef.PartitionKeys, error)
 	GetOld(context.Context) (typedef.PartitionKeys, error)
 	GiveOlds(context.Context, ...typedef.PartitionKeys)
-	ReleaseToken(uint32)
 }
 
 type Generator struct {
@@ -149,11 +148,6 @@ func (g *Generator) GiveOlds(ctx context.Context, tokens ...typedef.PartitionKey
 			g.oldDroppedValues.Inc()
 		}
 	}
-}
-
-// ReleaseToken removes the corresponding token from the in-flight tracking.
-func (g *Generator) ReleaseToken(token uint32) {
-	g.GetPartitionForToken(token).releaseToken(token)
 }
 
 func (g *Generator) start(ctx context.Context, wakeup chan struct{}) {

--- a/pkg/generators/partition.go
+++ b/pkg/generators/partition.go
@@ -16,6 +16,7 @@ package generators
 
 import (
 	"context"
+	"os"
 	"sync"
 	"sync/atomic"
 
@@ -27,8 +28,8 @@ import (
 
 type (
 	Partition struct {
-		values    chan typedef.PartitionKeys
-		oldValues chan typedef.PartitionKeys
+		values    atomic.Pointer[chan typedef.PartitionKeys]
+		oldValues atomic.Pointer[chan typedef.PartitionKeys]
 		wakeup    chan<- struct{}
 		isStale   atomic.Bool
 	}
@@ -61,8 +62,13 @@ func (p *Partition) get(ctx context.Context) (typedef.PartitionKeys, error) {
 // getOld returns a previously used value and token or a new if
 // the old queue is empty.
 func (p *Partition) getOld(ctx context.Context) (typedef.PartitionKeys, error) {
+	ch := p.oldValues.Load()
+	if ch == nil {
+		return typedef.PartitionKeys{}, os.ErrClosed
+	}
+
 	select {
-	case v := <-p.oldValues:
+	case v := <-*ch:
 		return v, nil
 	case <-ctx.Done():
 		return typedef.PartitionKeys{}, context.Canceled
@@ -75,8 +81,13 @@ func (p *Partition) getOld(ctx context.Context) (typedef.PartitionKeys, error) {
 // is empty, in which case it removes the corresponding token from the
 // in-flight tracking.
 func (p *Partition) giveOld(ctx context.Context, v typedef.PartitionKeys) bool {
+	ch := p.oldValues.Load()
+	if ch == nil {
+		return false
+	}
+
 	select {
-	case p.oldValues <- v:
+	case *ch <- v:
 		return true
 	case <-ctx.Done():
 		return false
@@ -86,17 +97,17 @@ func (p *Partition) giveOld(ctx context.Context, v typedef.PartitionKeys) bool {
 }
 
 func (p *Partition) push(v typedef.PartitionKeys) bool {
+	ch := p.values.Load()
+	if ch == nil {
+		return false
+	}
+
 	select {
-	case p.values <- v:
+	case *ch <- v:
 		return true
 	default:
 		return false
 	}
-}
-
-// releaseToken removes the corresponding token from the in-flight tracking.
-func (p *Partition) releaseToken(_ uint32) {
-	// p.inFlight.Delete(token)
 }
 
 func (p *Partition) wakeUp() {
@@ -107,9 +118,14 @@ func (p *Partition) wakeUp() {
 }
 
 func (p *Partition) pick(ctx context.Context) (typedef.PartitionKeys, error) {
+	ch := p.values.Load()
+	if ch == nil {
+		return typedef.PartitionKeys{}, os.ErrClosed
+	}
+
 	select {
-	case v := <-p.values:
-		if len(p.values) <= cap(p.values)/4 {
+	case v := <-*ch:
+		if len(*ch) <= cap(*ch)/4 {
 			p.wakeUp() // channel at 25% capacity, trigger generator
 		}
 		return v, nil
@@ -117,20 +133,26 @@ func (p *Partition) pick(ctx context.Context) (typedef.PartitionKeys, error) {
 		return typedef.PartitionKeys{}, context.Canceled
 	default:
 		p.wakeUp()
-		return <-p.values, nil
+		return <-*ch, nil
 	}
 }
 
 func (p *Partition) Close() error {
-	close(p.oldValues)
-	close(p.values)
+	ch := p.values.Swap(nil)
+	if ch != nil {
+		close(*ch)
+	}
+
+	oldCh := p.oldValues.Swap(nil)
+	if oldCh != nil {
+		close(*oldCh)
+	}
 
 	return nil
 }
 
 type Partitions struct {
 	parts []Partition
-	mu    sync.Mutex
 }
 
 var closePartitions sync.Once
@@ -138,9 +160,6 @@ var closePartitions sync.Once
 func (p *Partitions) Close() error {
 	var err error
 	closePartitions.Do(func() {
-		p.mu.Lock()
-		defer p.mu.Unlock()
-
 		for i := range len(p.parts) {
 			err = multierr.Append(err, p.parts[i].Close())
 		}
@@ -162,12 +181,12 @@ func NewPartitions(count, pkBufferSize int, wakeup chan<- struct{}) *Partitions 
 	partitions := make([]Partition, count)
 
 	for i := range len(partitions) {
-		partitions[i] = Partition{
-			values:    make(chan typedef.PartitionKeys, pkBufferSize),
-			oldValues: make(chan typedef.PartitionKeys, pkBufferSize),
-			wakeup:    wakeup,
-			isStale:   atomic.Bool{},
-		}
+		partitions[i] = Partition{wakeup: wakeup}
+
+		ch := make(chan typedef.PartitionKeys, pkBufferSize)
+		partitions[i].values.Store(&ch)
+		oldCh := make(chan typedef.PartitionKeys, pkBufferSize)
+		partitions[i].oldValues.Store(&oldCh)
 	}
 
 	return &Partitions{

--- a/pkg/jobs/mutation.go
+++ b/pkg/jobs/mutation.go
@@ -83,8 +83,6 @@ func (m *Mutation) run(ctx context.Context) error {
 		return err
 	}
 
-	defer m.generator.ReleaseToken(mutateStmt.PartitionKeys.Token)
-
 	err = m.store.Mutate(ctx, mutateStmt)
 
 	if err == nil {

